apiserver/facades/agent/provisioner/provisioner.go

```diff
@@ -987,21 +1024,16 @@ func (h *prepareOrGetHandler) ConfigType() string {
 func (h *prepareOrGetHandler) ProcessOneContainer(
 	ctx context.Context,
 	env environs.Environ, callContext envcontext.ProviderCallContext, policy BridgePolicy,
-	idx int, host, guest Machine, _ network.SubnetInfos,
+	idx int, host, guest Machine, hostInstanceID, guestInstanceID instance.Id, _ network.SubnetInfos,
 ) error {
-	instanceId, err := guest.InstanceId()
 	if h.maintain {
-		if err == nil {
+		if guestInstanceID != "" {
```

Comment:
I wonder if it would be better to give the instance.Id the ability to say if it was empty or not. Instead of forcing it into a string comparison. That way we can then define globally what empty truly is for a instance.Id.